apm-agent-plugins/apm-jdbc-plugin/src/main/java/co/elastic/apm/agent/jdbc/signature/Scanner.java

```diff
@@ -0,0 +1,404 @@
+/*-
+ * #%L
+ * Elastic APM Java agent
+ * %%
+ * Copyright (C) 2018 - 2019 Elastic and contributors
+ * %%
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * #L%
+ */
+package co.elastic.apm.agent.jdbc.signature;
+
+import javax.annotation.Nullable;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class Scanner {
+
+    @Nullable
+    private String input;
+    private int start; // text start pos in bytes
+    private int end; // text end pos in bytes
+    private int pos; // read pos in bytes
+
+    public void setQuery(String sql) {
+        this.input = sql;
+        start = 0;
+        end = 0;
+        pos = 0;
+    }
+
+    public Token scanWhile(Token token) {
+        for (Token t = scan(); t != Token.EOF; t = scan()) {
+            if (t != token) {
+                return t;
+            }
+        }
+        return Token.EOF;
+    }
+
+    public boolean scanUntil(Token token) {
+        for (Token t = scan(); t != Token.EOF; t = scan()) {
+            if (t == token) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    public boolean scanToken(Token token) {
+        for (Token t = scan(); t != Token.EOF; t = scan()) {
+            if (t == token) {
+                return true;
+            } else if (t != Token.COMMENT) {
+                return false;
+            }
+        }
+        return false;
+    }
+
+    public Token scan() {
+        if (!hasNext()) {
+            return Token.EOF;
+        }
+        char c = next();
+        while (Character.isSpaceChar(c) && hasNext()) {
+            c = next();
+        }
+        start = pos - 1;
```

Comment:
I just found a bug in the Go agent while reading your code :)
If you are dealing in bytes, then I think this should be `- number_of_bytes_in(c)` rather than `- 1`